### PR TITLE
Ham api

### PR DESCRIPTION
**Pull request by dodget**
This (draft) PR resolves [ATGU-2401](https://jira.huit.harvard.edu/browse/ATGU-2401) by:

- Adding a Harvard Art Museums api key configuration
- Adding a Harvard Art Museums archive to the `archives.rdf` file
- Adding `harvard_art_museums.xsl` style sheet to parse the content response from the api
- Adding a link to the archive in the dropdown selector

## Notes

- The naming of files of keys proved to be _very particular_. It seems that there was some magic involved in the association of api keys to their respective archives, and how the archive is found in `archives.rdf` (by the `dc:title` attribute).
- For this to work in production, the ops repo will also need to be updated with adding the api key to `userdata.sh` and all the rest of what needs doing (already done for scalar-dev)

### CODE REVIEW

**Comment by ColeDCrawford** [comment]
The importing seems to work pretty well as a first pass! 

One thing I noticed is that there are probably some improvements we can do around the metadata importing. If there’s an RDF field in which the manifest metadata and the entered data conflict (eg dcterms:date , the manifest wins (it’s processed later I think). We should be checking this and probably default to any data pulled by the API and/or entered by the user manually, rather than the metadata in the manifest. So if the field isn’t blank, use the API / user data instead of the manifest.

We can also look at pulling in additional metadata from the API - the fields I used in Tensor are between 237-268 here: [https://github.com/Harvard-ATG/tensor-profiles/blob/516a3c07a4ad3809a4a2cde866e2fee77c98bec3/parsers/harvard-art-museum/parser.js#L237-268](https://github.com/Harvard-ATG/tensor-profiles/blob/516a3c07a4ad3809a4a2cde866e2fee77c98bec3/parsers/harvard-art-museum/parser.js#L237-268)

The YouTube API is an example eg fields at [https://github.com/Harvard-ATG/scalar/blob/319aef92e094d42adba593478bdd5766a929401c/system/application/rdf/xsl/youtube_2.xsl#L32-L44](https://github.com/Harvard-ATG/scalar/blob/319aef92e094d42adba593478bdd5766a929401c/system/application/rdf/xsl/youtube_2.xsl#L32-L44)

I am also sometimes seeing an odd React error from Mirador, but I don't think this is necessarily related to the HAM importer. [See this test page](https://scalar-dev.fas.harvard.edu/cole-test-book-1/x-radiographs-of-still-life) for an example.

Screen Shot 2020-11-04 at 5 17 08 PM

**Comment by dodget** [comment]
Additional metadata fields have been added to the archive, and the iiif metadata import process does no longer overrides user-supplied inputs. The latest work has been deployed to scalar dev.

Note:
- There was an issue with deploying this work because `harvard_art_musems.xsl` was not executable. It was strange but ultimately resolved.

**Comment by ColeDCrawford** [comment]
Thanks @dodget - looking pretty good. Can confirm that IIIF metadata import process no longer overrides user-supplied metadata.

It does seem to be missing some fields from manifests, however. For instance, dcterms:date is not filled either from a IIIF manifest (metadata array, label:Date - https://iiif.harvardartmuseums.org/manifests/object/299843) or the HAM API (dated property https://api.harvardartmuseums.org/object/299843?apikey=addakeyhere). I think this is happening (at least to dcterms:date and a handful of other fields - I haven't done a comprehensive check) because Scalar prepopulates the metadata tab in the import wizard with "Featured Fields":
Screen Shot 2021-01-19 at 2 04 20 PM

Those fields are blank, but I'm guessing that the check for user-supplied metadata still sees them and defaults to using them instead of the HAM/manifest metadata. Scalar then saves the page to the database, but doesn't persist the blank field as RDF can't be blank. So the check for user metadata should only default to using that if the field is not blank. I think that should fix this ...

**Comment by dodget** [comment]
Ah thats an interesting case @ColeDCrawford - nice work catching that.

Just so we are on the same page, here is the list of the dcterms values that I am currently evoking from the HAM archive using xslt:
```
title
description
source
accrualMethod
available
medium
format
provenance
rights
rightsHolder
type		
spatial			
coverage
```

There were some terms, like date, that were proving complicated assemble into the formatted standard that dc terms expect. On top of that, I noticed the values we were getting from the api were not consistent enough to assemble in a prescribed way

File: system/application/rdf/xsl/harvard_art_museums.xsl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	<xsl:template match="*"></xsl:template>
+	
+	<xsl:template match="//node/node[images]">
+	    		
+		<rdf:Description rdf:about="{url}">
+		    <art:thumbnail rdf:resource="{concat(images/node[1]/iiifbaseuri, '/full/!225,225/0/default.jpg')}" />
+			<art:filename><xsl:value-of select="concat(seeAlso/node[1]/id, '?iiif-manifest=1')" /></art:filename>
+			<dcterms:title><xsl:value-of select="title" /></dcterms:title>
+			<dcterms:description><xsl:value-of select="description" /></dcterms:description>
+			<dcterms:source>Harvard Art Museums</dcterms:source>
```
Comment (arthurian) [inline system/application/rdf/xsl/harvard_art_museums.xsl]: Should this use the `archiveName` variable that's defined at the top of the XSL?

```xml
<dcterms:source><xsl:value-of select="$archiveName"/></dcterms:source>
```

File: system/application/rdf/xsl/harvard_art_museums.xsl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			<dcterms:provenance><xsl:value-of select="provenance" /></dcterms:provenance>
+			<dcterms:rights><xsl:value-of select="copyright" /></dcterms:rights>
+			<dcterms:rightsHolder><xsl:value-of select="creditline" /></dcterms:rightsHolder>
+			<dcterms:type><xsl:value-of select="classification" /></dcterms:type>
+			<xsl:if test="count(places/node/displayname) > 0">
+                <dcterms:spatial><xsl:value-of select="concat(places/node/displayname, ' (', places/node/type, ')')"/></dcterms:spatial>
+            </xsl:if>
+			<xsl:if test="count(culture) > 0">
+                <dcterms:coverage><xsl:value-of select="concat('Culture:', culture)" /></dcterms:coverage>
+            </xsl:if>
```
Comment (arthurian) [inline system/application/rdf/xsl/harvard_art_museums.xsl]: Minor nitpick: the indentation is a bit wonky around `<dcterms:spatial>` and `<dcterms:coverage>`, making it hard to read the conditional.

File: system/application/rdf/xsl/harvard_art_museums.xsl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			<xsl:apply-templates select="//node/node[images]" />
+		</rdf:RDF>		
+	</xsl:template>  
+
+	<xsl:template match="*"></xsl:template>
+	
+	<xsl:template match="//node/node[images]">
+	    		
+		<rdf:Description rdf:about="{url}">
+		    <art:thumbnail rdf:resource="{concat(images/node[1]/iiifbaseuri, '/full/!225,225/0/default.jpg')}" />
```
Comment (arthurian) [inline system/application/rdf/xsl/harvard_art_museums.xsl:31]: Nice use of the [!w,h size syntax](https://iiif.io/api/image/2.1/#size) for the thumbnail.

**Review by arthurian — APPROVED** [review]
@dodget Overall the code changes look good to me. I'm not that familiar with the Harvard Art Museum API or the import process so I don't have much insight to contribute there or on the metadata issues.

File: system/application/rdf/xsl/harvard_art_museums.xsl (hunk 2 above)
Comment (dodget) [reply system/application/rdf/xsl/harvard_art_museums.xsl]: Yeah, I often forget that scalar uses tabs not spaces

**Comment by ColeDCrawford** [comment]
Tested and ready to merge from my perspective.

**Pull request merged**